### ISSUE TITLE
Problems getting debian 9 system to work with uyuni  stable 4.0.2

### ISSUE THREAD

**Issue by gnuninu** [uyuni-project/uyuni#1356]
I have managed to register a debian 9 VM fine with Uyuni stable 4.0.2 but i have few problems with the debian repo assignment, they don't seem to be configured after the registration
From the UI under (debian) system events the repo tasks are still in a pending state

**OS**: openSUSE leap 15.1
**salt master:** salt 2019.2.0 (Fluorine) - RPM: salt-master-2019.2.0-lp151.5.3.1.x86_64
**salt minion**: salt-minion 2019.2.0 (Fluorine)  (from saltstack debian repo https://repo.saltstack.com/#debian)

I followed the steps below:

```
spacewalk-common-channels -u admin -a amd64-deb debian-9-pool-amd64
spacewalk-common-channels -u admin -a amd64-deb debian-9-amd64-main-security
spacewalk-common-channels -u admin -a amd64-deb debian-9-amd64-main-updates


spacewalk-repo-sync --channel debian-9-amd64-main-security &
spacewalk-repo-sync --channel debian-9-amd64-main-updates &
```

I have created a test/qa channel with all of the patches with:

`spacecmd softwarechannel_clonetree`

and then created an activation key - 1-qa-debian9-test - with the previously cloned qa channels attached to it.

i have added then saltstack debian 9 repo (python 3) on the debian client which includes salt 2019.02
https://repo.saltstack.com/#debian

The old repo are still present in `/etc/apt/sources.list` and while updating the apt cache it actually hangs when it comes to refresh the repo pushed by uyuni:

```
apt update

0% [Working]
(hanging)
```

Content of `/etc/apt/sources.list.d/susemanager\:channels.list `looks weird:

```
deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJucUJvTURRQ3RtdGZ3OFdtVUNZU0NRIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi1zZWN1cml0eSJdfQ.x1GzTTtyNzN422lPN9IYA1Du1IpASY6gRa0sKs_qj_g@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-security main

deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJ2d2I2YzhSZWZNRjVwOFM0cHFWZmhnIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi11cGRhdGVzIl19.iN6YmdMuinnNRRPBVhKUz86OVVwqM0QmzjUjvwwzQuA@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-updates main


deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJrX3FGMzFEV1F0WGphN1ZRa18xbWl3Iiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktcG9vbC1hbWQ2NCJdfQ.bgv2NAb-0WUHOfciGH_0B00y_j2CgnjVaSA_44wpet4@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-pool-amd64 main
```
 
On the salt master:

```
cat /usr/share/susemanager/salt/channels/channels.repo 
# Channels managed by SUSE Manager
# Do not edit this file, changes will be overwritten
#
{% for chan, args in pillar.get(pillar.get('_mgr_channels_items_name', 'channels'), {}).items() %}
{%- set protocol = salt['pillar.get']('pkg_download_point_protocol', 'https')%}
{%- set hostname = salt['pillar.get']('pkg_download_point_host', args['host'])%}
{%- set port = salt['pillar.get']('pkg_download_point_port', args.get('port', 443))%}
{%- if grains['os_family'] == 'Debian' %}
deb {{ '[trusted=yes]' if not pillar.get('mgr_metadata_signing_enabled', false) else '[signed-by=/usr/share/keyrings/mgr-archive-keyring.gpg]' }} https://{{ args['token'] }}@{{ args['host'] }}:{{ args.get('port', 443)}}/rhn/manager/download {{ chan }} main
{%- else %}
[{{ args['alias'] }}]
name={{ args['name'] }}
enabled={{ args['enabled'] }}
autorefresh={{ args['autorefresh'] }}
{%- if grains['os_family'] == 'RedHat' %}
baseurl={{protocol}}://{{hostname}}:{{port}}/rhn/manager/download/{{ chan }}
susemanager_token={{ args['token'] }}
{%- else %}
baseurl={{protocol}}://{{hostname}}:{{port}}/rhn/manager/download/{{ chan }}?{{ args['token'] }}
{%- endif %}
type={{ args['type'] }}
gpgcheck={{ args['gpgcheck'] }}
repo_gpgcheck={{ args['repo_gpgcheck'] }}
pkg_gpgcheck={{ args['pkg_gpgcheck'] }}
{%- endif %}

{% endfor %}
```

Can you help to debug?

**Comment by admd** [comment]
Hi  @gnuninu , do you still face the issue? Can you please share logs of rhn_web_ui and apache logs so we can check why is it hanging.

**Comment by gnuninu** [comment]
Hi @admd,

I have de-registered the system and tried the registration process one more time and I confirm i had the same behaviour as described above.
Please find attached the logs requested with the addition of the salt master logs.
The channels are in a pending state in the UI as shown in the picture pending_channels_tasks

as pointed previously the content of `/etc/apt/sources.list.d/susemanager\:channels.list` looks weird:

```
deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJucUJvTURRQ3RtdGZ3OFdtVUNZU0NRIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi1zZWN1cml0eSJdfQ.x1GzTTtyNzN422lPN9IYA1Du1IpASY6gRa0sKs_qj_g@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-security main

deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJ2d2I2YzhSZWZNRjVwOFM0cHFWZmhnIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi11cGRhdGVzIl19.iN6YmdMuinnNRRPBVhKUz86OVVwqM0QmzjUjvwwzQuA@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-updates main


deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1OTY4NDIxNzEsImlhdCI6MTU2NTMwNjE3MSwibmJmIjoxNTY1MzA2MDUxLCJqdGkiOiJrX3FGMzFEV1F0WGphN1ZRa18xbWl3Iiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktcG9vbC1hbWQ2NCJdfQ.bgv2NAb-0WUHOfciGH_0B00y_j2CgnjVaSA_44wpet4@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-pool-amd64 main
```

![pending_channels_tasks](https://user-images.githubusercontent.com/21009705/65731591-45be0580-e0be-11e9-9c98-be078fa57b7c.png)
[rhn_web_ui.log](https://github.com/uyuni-project/uyuni/files/3660256/rhn_web_ui.log)
[master.log](https://github.com/uyuni-project/uyuni/files/3660265/master.log)
[error.log](https://github.com/uyuni-project/uyuni/files/3660266/error.log)
[access.log](https://github.com/uyuni-project/uyuni/files/3660267/access.log)

**Comment by mateiw** [comment]
Hi @gnuninu. I'd suggest you:
- Delete the minion from the Uyuni server
- Uninstall `salt-minion` package from Saltstack upstream. Make sure to remove also the config under `/etc/salt`
- Add a repo for the `salt-minion` package from openSUSE https://download.opensuse.org/repositories/systemsmanagement:/saltstack:/products:/debian/Debian_9.0/. It has additional patches that might not be yet be present in the vanilla upstream package.
- Make sure you have the Debian upstream repos enabed in `/etc/apt/sources/list` and that no Uyuni repos are enabled.
- `apt install salt-minion`. Verify that you have installed the version coming from openSUSE.
- Bootstrap the minion via the UI
- Check if all events completed successfully. If so, all the APT sources from `/etc/apt/sources.list` should be disabled and the Uyuni repos should be enabled in `/etc/apt/sources.list.d/susemanager\:channels.list`.
- Try again to assign the channels to the minion.

The manual installation of the Salt minion is needed because currently `mgr-create-bootstrap-repo` doesn't support Debian, only Ubuntu.

> Content of /etc/apt/sources.list.d/susemanager\:channels.list looks weird:

For the minion to access a channel on the Uyuni server it needs to be authorized. For that we use an auth token in the repo url, hence the "weird" looking url.

**Comment by gnuninu** [comment]
Hi @mateiw 

thanks for your reply, following your suggestion i think i have made some progress let me recap what i did so far:

```
apt purge salt-minion

rm -rf /etc/salt/*

wget -O - https://download.opensuse.org/repositories/systemsmanagement:/saltstack:/products:/debian/Debian_9.0/Release.key | apt-key add -

vi /etc/apt/sources.list

deb https://download.opensuse.org/repositories/systemsmanagement:/saltstack:/products:/debian/Debian_9.0/ ./

apt update

apt install salt-minion

cat /etc/apt/sources.list.d/susemanager\:channels.list 
# Channels managed by SUSE Manager
# Do not edit this file, changes will be overwritten
#
```

subscribed manaully channels from system--> debian host --> software --> software channels
After this the content of the uyuni's source list changed:

```
cat /etc/apt/sources.list.d/susemanager\:channels.list 
# Channels managed by SUSE Manager
# Do not edit this file, changes will be overwritten
#

```
```
deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE2MDMxNDk2NTAsImlhdCI6MTU3MTYxMzY1MCwibmJmIjoxNTcxNjEzNTMwLCJqdGkiOiJoZTBEQW1WTm9jOEtiSEs4VTBjbUhRIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi1zZWN1cml0eSJdfQ.NMnCMQIa0Cons88oiM31cNP0rqItZrEXuMsX0lrvKN4@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-security main


deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE2MDMxNDk2NTAsImlhdCI6MTU3MTYxMzY1MCwibmJmIjoxNTcxNjEzNTMwLCJqdGkiOiIwN3F5V2NpZ0pDQkZUb3NwV3JLQkd3Iiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktcG9vbC1hbWQ2NCJdfQ.aDdLm70K0pAk-QbU6ZD5hv6llqEuVhzSSlQqtelDoxk@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-pool-amd64 main


deb [trusted=yes] https://eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE2MDMxNDk2NTAsImlhdCI6MTU3MTYxMzY1MCwibmJmIjoxNTcxNjEzNTMwLCJqdGkiOiJzMDRXN1lmLWdXbFFmQm5lRUlOekVRIiwib3JnIjoxLCJvbmx5Q2hhbm5lbHMiOlsicWEtZGViaWFuLTktYW1kNjQtbWFpbi11cGRhdGVzIl19.UE1jwh0moOF-IhNmqNF_MFY2c_tqAbb6gRkD90Zs8yM@uyuni4-osuse.home-lab.net:443/rhn/manager/download qa-debian-9-amd64-main-updates main

```

In the source.list file I can see the old repos have not been removed and are still commented out

yum update refreshes the uyuni repo fine but as you can see the old debian repo are also refreshed since they were not disabled during the bootstraping/registration phase

```
apt update
Ign:1 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates InRelease
Ign:2 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security InRelease
Ign:3 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 InRelease
Get:4 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates Release [763 B]
Get:5 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security Release [772 B]
Get:6 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 Release [715 B]
Ign:7 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates Release.gpg
Ign:8 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security Release.gpg
Ign:9 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 Release.gpg
Ign:10 http://mirror.positive-internet.com/debian stretch InRelease
Get:11 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates/main amd64 Packages [26.1 kB]
Get:12 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security/main amd64 Packages [652 kB]
Get:13 http://mirror.positive-internet.com/debian stretch-updates InRelease [91.0 kB]
Hit:14 https://download.opensuse.org/repositories/systemsmanagement:/saltstack:/products:/debian/Debian_9.0 ./ InRelease
Hit:15 http://mirror.positive-internet.com/debian stretch Release
Fetched 772 kB in 0s (888 kB/s)                     
Reading package lists... Done
```

**Comment by mateiw** [comment]
> Hi @mateiw
> 
> In the source.list file I can see the old repos have not been removed and are still commented out
> 
> yum update refreshes the uyuni repo fine but as you can see the old debian repo are also refreshed since they were not disabled during the bootstraping/registration phase
> 
> ```
> apt update
> Ign:1 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates InRelease
> Ign:2 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security InRelease
> Ign:3 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 InRelease
> Get:4 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates Release [763 B]
> Get:5 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security Release [772 B]
> Get:6 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 Release [715 B]
> Ign:7 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates Release.gpg
> Ign:8 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security Release.gpg
> Ign:9 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-pool-amd64 Release.gpg
> Ign:10 http://mirror.positive-internet.com/debian stretch InRelease
> Get:11 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-updates/main amd64 Packages [26.1 kB]
> Get:12 https://uyuni4-osuse.home-lab.net:443/rhn/manager/download test_debian-qa-debain9-debian-9-amd64-main-security/main amd64 Packages [652 kB]
> Get:13 http://mirror.positive-internet.com/debian stretch-updates InRelease [91.0 kB]
> Hit:14 https://download.opensuse.org/repositories/systemsmanagement:/saltstack:/products:/debian/Debian_9.0 ./ InRelease
> Hit:15 http://mirror.positive-internet.com/debian stretch Release
> Fetched 772 kB in 0s (888 kB/s)                     
> Reading package lists... Done
> ```

This is caused by a bug in Salt upstream. There's a [patch](https://github.com/saltstack/salt/pull/55257/files) for this but it's only included in the [openSUSE package](https://build.opensuse.org/package/show/systemsmanagement:saltstack:products:testing:debian/salt). 
If you can't use that package I'm afraid you would have to disable the default repos manually.

**Comment by gnuninu** [comment]
Hi,
Problems reported here are addressed by #PR1917

**Comment by gnuninu** [comment]
Problems reported here are addressed by PR #1917
